# ExoClick Technical Test


-


In [250]:
import pandas as pd
import numpy as np
import matplotlib as plt

In [251]:
#import data: 

df = pd.read_csv(r'C:\Users\pablo\Desktop\ExoClick\data_test2.csv')

In [252]:
#check how the data looks like:

df.head(10)

dattime  client country  ad_type  total_impressions  \
0  2021-06-06*06:00:00       1     EST       15              137.0   
1  2021-06-12*22:00:00       2     KEN       15             4424.0   
2  2021-06-05*20:00:00       1     NCL       15              271.0   
3  2021-06-23*17:00:00       1     EST       17                1.0   
4  2021-06-02*10:00:00       3     NCL       15            11701.0   
5  2021-06-11*16:00:00       4     KEN       15             1496.0   
6  2021-06-19*08:00:00       5     KEN       17               16.0   
7  2021-06-20*03:00:00       6     KEN       17                1.0   
8  2021-06-03*21:00:00       7     NCL       17                8.0   
9  2021-06-21*15:00:00       8     KEN       17                1.0   

    total_value  
0  1.370000e-01  
1  9.172066e+10  
2  2.710000e-01  
3  1.000000e-03  
4  4.871900e+00  
5  3.082000e-01  
6  1.600000e-01  
7  2.450000e-02  
8  1.840000e-01  
9  2.000000e-03

In [253]:
df.tail(10)

dattime  client country  ad_type  total_impressions  \
96213  2021-06-20*20:00:00      19     NCL       10              125.0   
96214  2021-06-21*16:00:00       2     KEN        2             3269.0   
96215  2021-06-01*04:00:00      17     KEN       10               10.0   
96216  2021-06-17*14:00:00      10     NCL       10              296.0   
96217  2021-06-19*23:00:00      24     ISL        2               14.0   
96218  2021-06-21*17:00:00      39     EST        2               15.0   
96219  2021-06-05*22:00:00       3     KEN       10              162.0   
96220  2021-06-26*02:00:00      23     ISL       10               11.0   
96221  2021-06-16*22:00:00      18     EST       17                6.0   
96222  2021-06-03*17:00:00       3     NCL       10            22200.0   

       total_value  
96213     0.625000  
96214    69.681281  
96215     0.050000  
96216     0.158500  
96217     0.070000  
96218     0.170939  
96219     0.099900  
96220     0.115894  
96221     0.035000  
96222    11.873900

##### Columns dictionary
- datetime: Date and hour in the format date*hour
- client: Client identifier
- country: ISO3 country code
- ad_type: Ad type identifier
- impressions: Number of ad impressions
- value: Total value in cents paid by the client for those impressions

## Exploratory Data Analysis

First look at the dataset and some descriptors


In [254]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96223 entries, 0 to 96222
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   dattime            96223 non-null  object 
 1   client             96223 non-null  int64  
 2   country            96223 non-null  object 
 3   ad_type            96223 non-null  int64  
 4   total_impressions  96223 non-null  float64
 5   total_value        96223 non-null  float64
dtypes: float64(2), int64(2), object(2)
memory usage: 4.4+ MB


In [255]:
df.describe()

client       ad_type  total_impressions   total_value
count  96223.000000  96223.000000       96223.000000  9.622300e+04
mean      16.665174     11.487773        4045.099815  9.532133e+05
std       14.327396      5.057308       31901.013999  2.956839e+08
min        1.000000      2.000000      -19634.000000  0.000000e+00
25%        4.000000     10.000000           7.000000  2.010000e-02
50%       13.000000     10.000000          42.000000  1.400000e-01
75%       25.000000     17.000000         378.000000  1.104459e+00
max       73.000000     17.000000      941289.000000  9.172066e+10

In [256]:
df['client'].unique() #73 clients

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 53, 58,
       50, 52, 57, 55, 62, 59, 51, 63, 54, 64, 65, 66, 60, 56, 67, 68, 69,
       70, 61, 71, 72, 73], dtype=int64)

In [257]:
df['country'].unique() #6 countries - **ISL is repeated and in NCL there is a duplicate string**

array(['EST', 'KEN', 'NCL', 'ISL', 'ISL1', 'NCLNCL'], dtype=object)

In [258]:
df['ad_type'].unique() #4 types of ads

array([15, 17, 10,  2], dtype=int64)

### Data malfunctions found in EDA

- `Dattime` has to be changed to datetime format
- There are negative values of `total impressions`
- Some variables are not in a correct format
    - `ad type` and `client` should be an object
    - `total impressions` should be an integer

## Data cleaning

### 1. Missing values

In [259]:
#check for missing values

df.isnull().any()

#no missing values

dattime              False
client               False
country              False
ad_type              False
total_impressions    False
total_value          False
dtype: bool

### 2. Date time format

In [260]:
df['dattime'] #format='%Y-%m-%d*%H:%M:%S'

0        2021-06-06*06:00:00
1        2021-06-12*22:00:00
2        2021-06-05*20:00:00
3        2021-06-23*17:00:00
4        2021-06-02*10:00:00
                ...         
96218    2021-06-21*17:00:00
96219    2021-06-05*22:00:00
96220    2021-06-26*02:00:00
96221    2021-06-16*22:00:00
96222    2021-06-03*17:00:00
Name: dattime, Length: 96223, dtype: object

In [261]:
#transform date time in pandas format

df['dattime'] = pd.to_datetime(df['dattime'], format='%Y-%m-%d*%H:%M:%S')

### 3. Check for negative values in impressions 

In [262]:
df[df['total_impressions'] < 0]

dattime  client country  ad_type  total_impressions  \
87 2021-06-15 03:00:00       4     EST       15           -19634.0   

    total_value  
87       8.4741

In [263]:
# check for similar entries to decide if it is possible to delete the row

df[(df['client']==4) & (df['ad_type']==15) & (df['country']=='EST')].sort_values(by='dattime')

dattime  client country  ad_type  total_impressions  \
38269 2021-06-01 00:00:00       4     EST       15            11791.0   
25330 2021-06-01 02:00:00       4     EST       15            11885.0   
19874 2021-06-01 06:00:00       4     EST       15            14596.0   
10596 2021-06-01 10:00:00       4     EST       15            16304.0   
1180  2021-06-01 16:00:00       4     EST       15            24289.0   
...                   ...     ...     ...      ...                ...   
11273 2021-06-29 18:00:00       4     EST       15            10119.0   
22580 2021-06-29 19:00:00       4     EST       15             6016.0   
26352 2021-06-29 21:00:00       4     EST       15             3261.0   
20184 2021-06-29 22:00:00       4     EST       15             3153.0   
36294 2021-06-29 23:00:00       4     EST       15             4309.0   

       total_value  
38269       5.5461  
25330       6.4113  
19874       7.1502  
10596       8.1966  
1180       11.5780  
...            ...  
11273       2.7341  
22580       1.6207  
26352       0.8267  
20184       0.7921  
36294       1.1176  

[331 rows x 6 columns]

In [264]:
#remove the column with negative values since it is just 1 entry

df = df.drop(df[df['total_impressions'] < 0].index)

In [265]:
df[df['total_impressions'] < 0] 

#no negative impressions

Empty DataFrame
Columns: [dattime, client, country, ad_type, total_impressions, total_value]
Index: []

### 4. Check for duplicates 

In [266]:
#observe duplicates

df[df.duplicated()]

dattime  client country  ad_type  total_impressions  \
22173 2021-06-01 16:00:00      12     KEN       17                1.0   
22816 2021-06-14 01:00:00       1     KEN       17                2.0   
25858 2021-06-20 05:00:00      52     ISL       17                1.0   
26437 2021-06-24 08:00:00      36     KEN       17                2.0   
26813 2021-06-01 20:00:00      19     EST       17               10.0   
33350 2021-06-19 16:00:00       1     KEN       10                5.0   
37827 2021-06-11 22:00:00       1     EST       10               12.0   
47602 2021-06-18 06:00:00      10     ISL       17                1.0   
50307 2021-06-10 01:00:00       1     EST       15               17.0   
56866 2021-06-15 05:00:00       1     ISL       15                1.0   
57745 2021-06-14 20:00:00       1     KEN       15                1.0   
58047 2021-06-29 11:00:00       1     ISL       10               25.0   
60644 2021-06-29 00:00:00       1     KEN       10                2.0   
63445 2021-06-11 09:00:00      12     EST       17                6.0   
64127 2021-06-16 23:00:00       1     KEN       15                1.0   
71627 2021-06-11 04:00:00       1     EST       10               59.0   
74490 2021-06-27 07:00:00      36     KEN       17                1.0   
80971 2021-06-28 01:00:00       1     EST       10                1.0   
88873 2021-06-19 11:00:00       1     ISL       15                2.0   
89508 2021-06-23 15:00:00       1     KEN       10               20.0   
92093 2021-06-11 23:00:00       1     NCL       15               64.0   
93903 2021-06-21 19:00:00      28     NCL       10                1.0   

       total_value  
22173       0.0050  
22816       0.0020  
25858       0.0050  
26437       0.0100  
26813       0.0500  
33350       0.0050  
37827       0.0120  
47602       0.0005  
50307       0.0170  
56866       0.0010  
57745       0.0010  
58047       0.0250  
60644       0.0020  
63445       0.0180  
64127       0.0010  
71627       0.0590  
74490       0.0050  
80971       0.0010  
88873       0.0020  
89508       0.0200  
92093       0.0640  
93903       0.0100

In [267]:
print(f"Number of duplicate rows: {df.duplicated().sum()}")

Number of duplicate rows: 22


In [268]:
#remove duplicates since those are exact row duplicates

df = df.drop_duplicates()

### 5. Quality of the data

In [269]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 96200 entries, 0 to 96222
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   dattime            96200 non-null  datetime64[ns]
 1   client             96200 non-null  int64         
 2   country            96200 non-null  object        
 3   ad_type            96200 non-null  int64         
 4   total_impressions  96200 non-null  float64       
 5   total_value        96200 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(2), object(1)
memory usage: 5.1+ MB


In [ ]:
# change total_impressions to a int64 since it represents a discrete value
#fist, check if there are any impression with decimals:
df[df['total_impressions'] % 1 != 0]


dattime  client country  ad_type  total_impressions  \
4005 2021-06-16 04:00:00       4     ISL       15             4502.5   

      total_value  
4005       5.9103

In [271]:
#I decide to remove the row since it is just 1 

df = df.drop(df[df['total_impressions'] % 1 != 0].index)

In [272]:
#now transform total impressions into int65:

df['total_impressions'] = df['total_impressions'].astype('Int64')

print(df['total_impressions'].dtype)

Int64


##### Data Quality Issue: Decimal Impressions
- Found 1 row with non-integer impression count (value: [ 4502.5])
- Represents ~0.001% of the dataset (1 out of [total rows])
- Decision: Rounded down using np.floor() because:
  - Only 1 row affected (negligible impact on analysis)
  - Impressions must be whole numbers (can't have fractional ad views)
  - Floor function ensures conservative estimate
- Converted column from float64 to int64 after correction

In [280]:
#transform 'client' and 'ad_type' to an object:

df['client'] = df['client'].astype('category')

df['ad_type'] = df['ad_type'].astype('category')


In [281]:
#Check for the uniqueness in country:

df['country'].unique()

array(['EST', 'KEN', 'NCL', 'ISL'], dtype=object)

- ISL its duplicated with a 1 on it

- NCL has an entry with a duplicate (NCL, NCL)


In [275]:
df[df['country'] == 'ISL1']

dattime client country ad_type  total_impressions  \
4397 2021-06-02 15:00:00     21    ISL1      17                  6   

      total_value  
4397     0.031105

In [276]:
df[df['country'] == 'NCLNCL']

dattime client country ad_type  total_impressions  \
89382 2021-06-17 03:00:00     12  NCLNCL      10                 13   

       total_value  
89382        0.037

In [277]:
df['country'] = df['country'].replace('ISL1', 'ISL')
df['country'] = df['country'].replace('NCLNCL', 'NCL')

In [278]:
df['country'].unique()

array(['EST', 'KEN', 'NCL', 'ISL'], dtype=object)

In [282]:
#final check of the data:

df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 96199 entries, 0 to 96222
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   dattime            96199 non-null  datetime64[ns]
 1   client             96199 non-null  category      
 2   country            96199 non-null  object        
 3   ad_type            96199 non-null  category      
 4   total_impressions  96199 non-null  Int64         
 5   total_value        96199 non-null  float64       
dtypes: Int64(1), category(2), datetime64[ns](1), float64(1), object(1)
memory usage: 3.9+ MB


## Features 


- Add a new column with the eCPM - effective Cost Per Mille

In [301]:
def calculate_ecpm(row):
    '''
    Calculate eCPM (effective Cost Per Mille).
    
    eCPM = (total_value / total_impressions) * 1000
    
    Parameters:
    -----------
    row : pandas.Series
        A row from the DataFrame containing 'total_value' and 'total_impressions'
    
    Returns:
    --------
    float
        eCPM value in cents per thousand impressions, or 0 if impressions = 0'''
    
    if row['total_impressions'] > 0:
        return (row['total_value'] / row['total_impressions']) * 1000
    else:
        return 0

In [302]:
df['eCPM'] = df.apply(calculate_ecpm, axis=1)

dattime client country ad_type  total_impressions  \
14    2021-06-05 12:00:00     13     EST      17                197   
57    2021-06-25 09:00:00     13     ISL      17                 34   
88    2021-06-26 00:00:00     13     NCL      17                 96   
100   2021-06-06 01:00:00     13     ISL      17                 41   
138   2021-06-01 19:00:00     13     EST      17                119   
...                   ...    ...     ...     ...                ...   
96067 2021-06-18 05:00:00     13     ISL      10                 51   
96070 2021-06-18 05:00:00     13     EST      10                171   
96109 2021-06-01 17:00:00     13     EST      10                269   
96177 2021-06-15 04:00:00     13     ISL      10                 38   
96197 2021-06-29 21:00:00     13     EST      10                 29   

       total_value  eCPM  
14             0.0   0.0  
57             0.0   0.0  
88             0.0   0.0  
100            0.0   0.0  
138            0.0   0.0  
...            ...   ...  
96067          0.0   0.0  
96070          0.0   0.0  
96109          0.0   0.0  
96177          0.0   0.0  
96197          0.0   0.0  

[2770 rows x 7 columns]

## Univariate statistics 
